### PR TITLE
[ALLUXIO-2174] Add unit testing guide

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-2174

image

### CODE REVIEW

File: docs/en/Unit-Testing.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+---
+layout: global
+title: Unit Testing
```
Comment (gpang) [inline docs/en/Unit-Testing.md]: maybe add "Guidelines"?

File: docs/en/Unit-Testing.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+nickname: Unit Testing
+group: Resources
+---
+
+* Table of Contents
+{:toc}
+
+# Unit testing goals
+1. Unit tests act as examples of how to use the code under test.
```
Comment (gpang) [inline docs/en/Unit-Testing.md]: why are these starting from 1, 2, 3, but the next section starts from 0, 1, 2?

File: docs/en/Unit-Testing.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+2. Unit tests detect when an object breaks it's specification.
+3. Unit tests *don't* break when an object is refactored but still meets the same specification.
+
+# How to write a unit test
+
+0. If creating an instance of the class takes some work, create `@Before` and `@After` methods which handle setup and cleanup. Only do general setup which will apply to every test. Test-specific setup should be done locally in the tests that need it.
```
Comment (gpang) [inline docs/en/Unit-Testing.md]: Add a bit of explanation of the following example.

File: docs/en/Unit-Testing.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	      mMaster = new BlockMaster(blockJournal, mClock, mExecutorService);
+	      mMaster.start(true);
+	    }
+
+	    @After
+	    public void after() throws Exception {
+	      mMaster.stop();
+	    }
+
+1. Decide on an element of functionality to test. The functionality you decide to test should be part of the public API and should not care about implementation details.
```
Comment (gpang) [inline docs/en/Unit-Testing.md]: Any guidelines on what to test? Isn't the goal to test all public functionality?

File: docs/en/Unit-Testing.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+2. Give your test a name that describes what it’s testing. The functionality being tested should be simple enough to describe this way:
```
Comment (gpang) [inline docs/en/Unit-Testing.md]: Any guidelines on how to name the test?

**Comment by gpang** [comment]
@aaudiber Thanks! I left a few comments.

File: docs/en/Unit-Testing.md (hunk 5 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Added some examples. The guideline is that the test should be named after the functionality that it's testing.

File: docs/en/Unit-Testing.md (hunk 4 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Added a last step which loops back to here

File: docs/en/Unit-Testing.md (hunk 3 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Done

File: docs/en/Unit-Testing.md (hunk 2 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: fixed to index from 1

File: docs/en/Unit-Testing.md (hunk 1 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Done

**Comment by aaudiber** [comment]
@gpang Thanks for the review! I've addressed the comments, PTAL

**Comment by gpang** [comment]
LGTM

**Comment by ifcharming** [comment]
LGTM

File: docs/en/Unit-Testing.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+# Unit testing goals
+1. Unit tests act as examples of how to use the code under test.
+2. Unit tests detect when an object breaks it's specification.
+3. Unit tests *don't* break when an object is refactored but still meets the same specification.
+
+# How to write a unit test
+
+1. If creating an instance of the class takes some work, create `@Before` and `@After` methods which handle setup and cleanup. These methods get run automatically before and after each unit test. Only do general setup which will apply to every test. Test-specific setup should be done locally in the tests that need it. In this example, we are testing a `BlockMaster`, which depends on a journal, clock, and executor service. The executor service and journal we provide are real implementations, and the `TestClock` is a fake clock which can be controlled by unit tests.
+
+	    @Before
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: NIT: put <pre>```java</pre> to have syntax highlight?

File: docs/en/Unit-Testing.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	      mMaster = new BlockMaster(blockJournal, mClock, mExecutorService);
+	      mMaster.start(true);
+	    }
+
+	    @After
+	    public void after() throws Exception {
+	      mMaster.stop();
+	    }
+
+2. Decide on an element of functionality to test. The functionality you decide to test should be part of the public API and should not care about implementation details.
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: also mentioning that each test case should be simple and better focus testing one thing.

File: docs/en/Unit-Testing.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+{:toc}
+
+# Unit testing goals
+1. Unit tests act as examples of how to use the code under test.
+2. Unit tests detect when an object breaks it's specification.
+3. Unit tests *don't* break when an object is refactored but still meets the same specification.
+
+# How to write a unit test
+
+1. If creating an instance of the class takes some work, create `@Before` and `@After` methods which handle setup and cleanup. These methods get run automatically before and after each unit test. Only do general setup which will apply to every test. Test-specific setup should be done locally in the tests that need it. In this example, we are testing a `BlockMaster`, which depends on a journal, clock, and executor service. The executor service and journal we provide are real implementations, and the `TestClock` is a fake clock which can be controlled by unit tests.
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: do we allow  `@BeforeClass` and `@AfterClass` in unit tests?

File: docs/en/Unit-Testing.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+	    @After
+	    public void after() throws Exception {
+	      mMaster.stop();
+	    }
+
+2. Decide on an element of functionality to test. The functionality you decide to test should be part of the public API and should not care about implementation details.
+3. Give your test a name that describes what functionality it's testing. The functionality being tested should ideally be simple enough to fit into a name, e.g. `removeNonexistentBlockThrowsException`, `mkdirCreatesDirectory`, or `cannotMkdirExistingFile`.
+
+	    @Test
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: shall we java-doc this test case?

File: docs/en/Unit-Testing.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	    public void before() throws Exception {
+	      Journal blockJournal = new ReadWriteJournal(mTestFolder.newFolder().getAbsolutePath());
+	      mClock = new TestClock();
+	      mExecutorService =
+	          Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("TestBlockMaster-%d", true));
+	      mMaster = new BlockMaster(blockJournal, mClock, mExecutorService);
+	      mMaster.start(true);
+	    }
+
+	    @After
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: maybe it worths a separate bullet to emphasize and illustrate that resource needs to be released.

File: docs/en/Unit-Testing.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+layout: global
+title: Unit Testing Guidelines
+nickname: Unit Testing Guidelines
+group: Resources
+---
+
+* Table of Contents
+{:toc}
+
+# Unit testing goals
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: do we want to cover integration tests too? or it will be in a separate article?

File: docs/en/Unit-Testing.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+# Conventions
+1. The tests for `src/main/java/ClassName.java` should go in `src/test/java/ClassNameTest.java`
+2. Tests do not need to handle or document specific checked exceptions. Prefer to simply add `throws Exception` to the test method signature.
+3. Aim to keep tests short and simple enough that they don't require comments to understand.
+
+# Patterns to avoid
+
+1. Avoid randomness. Edge cases should be handled explicitly.
+2. Avoid waiting for something by calling Thread.sleep(). This leads to slower unit tests and can cause flaky failures if the sleep isn't long enough.
+3. Avoid modifying global state. We typically use JUnit `Rule` when this is necessary so that the state is properly restored. See `SystemPropertyRule`, `TtlIntervalRule`, or `LocalAlluxioClusterResource` for examples of this.
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: Needs more details explaining these two rules and resource

File: docs/en/Unit-Testing.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+4. Avoid using Whitebox to mess with the internal state of objects under test. If you need to mock a dependency, change the object to take the dependency as a parameter in its constructor (see [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection))
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: is it possible for us to  also avoid powermock?

File: docs/en/Unit-Testing.md (hunk 9 above)
Comment (apc999) [inline docs/en/Unit-Testing.md]: also `throws Exception` is probably recommended / required?

File: docs/en/Unit-Testing.md (hunk 8 above)
Comment (apc999) [inline docs/en/Unit-Testing.md]: How about making the example in quote mode?

File: docs/en/Unit-Testing.md (hunk 11 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: that would go in a separate article.

File: docs/en/Unit-Testing.md (hunk 8 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: I wouldn't encourage `@BeforeClass`/`@AfterClass` since it makes it easier for tests to interfere with each other, but I also wouldn't want to ban it outright since it can be useful when used right.

File: docs/en/Unit-Testing.md (hunk 6 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Good call, added java syntax highlighting

File: docs/en/Unit-Testing.md (hunk 10 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Added

File: docs/en/Unit-Testing.md (hunk 7 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Added

File: docs/en/Unit-Testing.md (hunk 9 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: I don't think javadoc here is worth the cost in space and maintenance.

File: docs/en/Unit-Testing.md (hunk 12 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Added what they do

File: docs/en/Unit-Testing.md (hunk 13 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Our codebase is still too full of singletons and final classes which don't implement interfacess

**Comment by aaudiber** [comment]
Thanks for all the great feedback @apc999! I've addressed the comments, please take another look when you have a chance.

**Comment by apc999** [comment]
LGTM. thanks for the improvement!
